tiled/server/router.py

```diff
@@ -1215,3 +1219,104 @@ async def delete_revision(
 
     await entry.delete_revision(number)
     return json_or_msgpack(request, None)
+
+
+@router.get("/asset/bytes/{path:path}")
+async def get_asset(
+    request: Request,
+    id: int,
+    relative_path: Optional[Path] = None,
+    entry=SecureEntry(scopes=["read:data"]),  # TODO: Separate scope for assets?
+):
+    if not hasattr(entry, "asset_by_id"):
+        raise HTTPException(
+            status_code=405,
+            detail="This node does not support downloading assets.",
+        )
+
+    asset = await entry.asset_by_id(id)
+    if asset is None:
+        raise HTTPException(
+            status_code=404,
+            detail=f"This node exists but it does not have an Asset with id {id}",
+        )
+    if asset.is_directory:
+        if relative_path is None:
```

Comment:
Good question.

A directory is not something you can put into an HTTP response. You would have to read and combine all the files somehow into a single byte stream.

A reasonable way to do this would be to ZIP the directory, similarly to what happens when you download a directory from Google Drive. But this opens up a bunch more complexity for the server.

Zipping up a big directory might take longer than the HTTP client (or proxy) timeout internal, so the request would need to launch an asynchronous job, where the user is given a link that they can use to check on the status and eventually download the result. And the server would need some storage area to stage the ZIP archive for later download, with rules about how long to keep a given result and how big they are allowed to be.

This could all be done, but it seems simpler for the moment to do one file per request, spooled from disk using `FileResponse`, and let the client piece it together to make things simple for the user. The client could even ZIP it if desired.

Are we missing a simpler way to handle this?

Reply:
Having said all that….it might be possible to incrementally construct and send a ZIP in chunks, reading a chunk of each file, appending it to the ZIP, and then draining the ZIP-in-progress into the socket. In this way, similar to `FileResponse` the  server memory usage is bounded, regardless of file or directory size.

I’ll look at this. If it works, then maybe we should accept `/asset/bytes/{path}?id=…` with `relative_path` optional, even for directory-backed assets, and stream a zip of the whole directory if no `relative_path` is given.

I think it would still be worth keeping the option of specifying `relative_path` for selective/incremental access to directory-based assets.

Reply:
This doesn't quite work, which I suppose is why Google Drive _et. al._ don't support it either. This [failed attempt](https://gist.github.com/danielballan/20dfb2b41902cc9396a13f434ca15045) illustrates the idea in my last comment as code.

ZIP places a header before the bytes from each file, and the header contains a hash (CRC-32) of the uncompressed file. Thus, the server needs to read the whole file in order to send the hash before it begins to sends the file's contents. That leaves us with two bad options:

1. Read a file fully into memory once, hash it, and send it.
2. Stream through each file in chunks, twice. The first time, compute the hash. Send the file header. The second time, send the contents in chunks.

(1) puts too much memory pressure on the server, and files may be very large. (2) requires reading the file from disk twice.

So I think we're back to, "There is no simple way to stream a directory over HTTP."

Reply:
That's fine, I think I misunderstood tiled/client/base `raw_export` then- presumably that will never support >1 data_source?

Reply:
I think it could. `raw_export` takes a directory as its argument (defaults to `cwd`) and can put 1 or N files in there, with names provided by the server. Today, this can support use cases like:

- one HDF5 file (1 data source, 1 asset, 1 file)
- a Zarr dataset (1 data source, 1 asset, 1 directory of N files)
- an HDF5 with virtual datasets (1 data source, N assets, N files)
- a sequence of TIFFs (1 data source, N assets, N files)

All of the above are 1 data source. How would we ever end up with N data sources? A Bluesky stream can be backed by an HDF5 for image data and some tabular format (Arrow, or even CSV, anything...) for time series tabular data.

The SQL schema supports a one-to-many relation between logical nodes and data sources. The goal is to present the user with one logical entity, which just happens to be backed by a couple different data sources, and the user doesn't have to care. But both the server and client code have `NotImplementedErrors` sprinkled into these code paths for now.

I _think_ `raw_export` can be extended to deal with this when the time comes, since it already supports downloading N files.

Reply:
I'll mark this as resolved, then. Glad we explored the question and understand the limitations more specifically.